File: wi-seong-cheol/BOJ/0x05 Stack/3015.cpp
//
//  3015.cpp
//  main
//
//  Created by wi_seong on 2022/11/07.
//

/*
 [Input]
 7
 2
 4
 1
 2
 2
 5
 1
 [Output]
 10
 */
// 시간복잡도: O(n)
// 최악시간: 500,000
// 난이도: Gold 1
// Timer: 19m
// Url: https://www.acmicpc.net/problem/3015

#include <iostream>
#include <stack>

using namespace std;
#define X first
#define Y second
#define ll long long
int n;
ll ans;
stack<pair<int, int>> S;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    cin >> n;
    while(n--) {
        int x; cin >> x;
        int cnt = 1;
        while(!S.empty() && S.top().X <= x) {
            ans += S.top().Y;
            if(S.top().X == x) cnt += S.top().Y;
            S.pop();
        }
        if(!S.empty()) ans++;
        S.push({x, cnt});
    }
    cout << ans;
    
    return 0;
}

File: hall_of_fame/wi-seong-cheol/BOJ/0x05 Stack I/3015.swift
//
//  3015.swift
//  wi-seong-cheol
//
//  Created by wi_seong on 2022/12/14.
//

/*
 [Input]
 7
 2
 4
 1
 2
 2
 5
 1
 [Output]
 10
 */
// 시간복잡도: O(n)
// 최악시간: 500,000
// 난이도: Platinum 1
// Timer: 19m
// Url: https://www.acmicpc.net/problem/3015

import Foundation

// 입력
let n = Int(readLine()!)!
var stack: [(Int, Int)] = []
var result = 0

// 풀이
for _ in 0 ..< n {
    let tall = Int(readLine()!)!
    var cnt = 1
    while !stack.isEmpty, stack.last!.0 <= tall {
        result += stack.last!.1
        if stack.last!.0 == tall {
            cnt += stack.last!.1
        }
        stack.removeLast()
    }
    if(!stack.isEmpty) {
        result += 1
    }
    stack.append((tall, cnt))
}

// 출력
print(result)
